cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt

```diff
@@ -273,6 +356,60 @@ abstract class Language<T : LanguageFrontend<*, *>> : Node() {
             signature.size == targetSignature.size
         }
     }
+
+    /**
+     * This functions gives the language a chance to refine the results of a
+     * [ScopeManager.resolveCall]. The default implementation will follow the following heuristic:
+     * - If the list of viableFunctions is empty, we can directly return.
+     * - If we have only one item in viableFunctions, we can take it.
+     * - Next, we can check for direct matches, meaning that they have a SignatureResult that only
+     *   has DirectMatchNoCastNeeded casts.
+     * - Lastly, if we have not direct matches, we need to sort the viable functions using a simple
+     *   ranking. The function(s) will the best (lowest) ranking will be chosen as the best. The
+     *   ranking is determined by the "depth" of all cast results in the signature results.
+     */
+    open fun bestViableResolution(
+        result: ScopeManager.CallResolutionResult
+    ): List<FunctionDeclaration> {
+        // No need to do anything, if the list of viable functions is empty
+        if (result.viableFunctions.isEmpty()) {
+            return listOf()
+        }
+
+        // If we already have only one viable result, we can just go back
+        val single = result.viableFunctions.singleOrNull()
+        if (single != null) {
+            return listOf(single)
+        }
+
+        // Check for direct matches. Let's hope there is only one, otherwise we have an ambiguous
+        // result
+        val directMatches = result.signatureResults.entries.filter { it.value.isDirectMatch }
+        return if (directMatches.size > 1) {
+            // This is an ambiguous result
+            result.type = ScopeManager.CallResolutionResult.Type.AMBIGUOUS
```

Comment:
This should be fine if we have languages where we can overload or override functions/methods. Otherwise it would be PROBLEMATIC (e.g. in C), right? Can we use our LanguageTraits here to make this choice?

Reply:
We probably need to differentiate between overloading and overriding. Even if you overload a function, there can be a point, in which we cannot differentiate between two function candidates, which are equally ranked and thus ambiguous. A compiler will return an error. I described this now in the doc.

I am not 100 % sure how we will deal with overridden functions, in this case they could still both match but the result is probably "ok". Not sure if we will only encounter this once I try to implement method resolution.

Reply:
I agree that we have to differentiate between these two issues and that it's probably a future topic.